### DISCUSSION TITLE
How to pin package to specific commit

### DISCUSSION THREAD

**Discussion by emacsbliss** [SystemCrafters/crafted-emacs#397] [Q&A] [answered]
similar to what DOOM Emacs does, in DOOM Emacs, we can use `:pin` to specify the commit hash for a package. Does crafted-emacs has something similar? didn't find it so far.

**Comment by jvdydev** [comment]
Hi,

Crafted Emacs uses the built-in `package.el` package manager (by default) which does not support pinning packages (at least not to my knowledge).
Doom Emacs uses a third party package manager (`straight.el`) which has this functionality (which Doom Emacs re-exposes within its own system for package management).

You *can* use `straight.el` (or other third party package managers) with Crafted Emacs and use the additional functionality that way (see [using alternative package managers with Crafted Emacs](https://github.com/SystemCrafters/crafted-emacs/blob/master/docs/crafted-emacs.org#using-alternate-package-managers)), but Crafted Emacs does not provide any additional facilities for managing packages itself.

**Reply by emacsbliss** [reply]
ok, thanks a lot!

**Comment by jeffbowman** [comment, marked as answer]
Pinning packages is possible, simply update the variable `package-load-list` (see the info page for complete information). 

Here is an example:

```emacs-lisp
(customize-set-variable 'package-load-list
                        '((doom-modeline nil)          ; disallows doom-modeline, if it is installed, to ever be loaded.
                          (tree-sitter-langs "0.12.8") ; only allows tree-sitter-langs 0.12.8 to be loaded regardless if other versions are installed.
                          all))                                    ; load everything else normally using the most current version installed.
```

This method does *not* prevent the user from installing different versions of the same package, it only picks the one to load and activate when Emacs is starting.  This example came from a time when loading tree-sitter-langs (for Emacs 28) was broken in versions after 0.12.8.  The fix was resolved several versions later.  Also, there was a time when doom modeline was installed by default in Crafted Emacs, however, due to the number of problems it caused, we stopped automatically installing it.  In my case I had it installed, but never wanted to use it, so setting the version to `nil` prevented it from loading or activating even though I had it installed.

Hope that helps!

cc: @emacsbliss @jvdydev

**Reply by jvdydev** [reply]
my bad, learned something new :smile: (also no pinning to a specific commit, but that's more due to the nature of `package.el` using versioned packages instead of git repositories).

**Reply by jeffbowman** [reply]
Yep.  Although, that may be possible with `package-vc` but I haven't looked at that in a while.
